sys/event/thread.c

```diff
@@ -78,48 +89,37 @@ void event_thread_init(event_queue_t *queue, char *stack, size_t stack_size,
 #define EVENT_THREAD_LOWEST_PRIO   (THREAD_PRIORITY_IDLE - 1)
 #endif
 
-#ifdef MODULE_EVENT_THREAD_HIGHEST
-event_queue_t event_queue_highest;
-static char _evq_highest_stack[EVENT_THREAD_HIGHEST_STACKSIZE];
-#endif
-
-#ifdef MODULE_EVENT_THREAD_MEDIUM
-event_queue_t event_queue_medium;
-static char _evq_medium_stack[EVENT_THREAD_MEDIUM_STACKSIZE];
-#endif
+/* rely on compiler / linker to garbage collect unused stacks */
+static char WORD_ALIGNED _evq_highest_stack[EVENT_THREAD_HIGHEST_STACKSIZE];
+static char WORD_ALIGNED _evq_medium_stack[EVENT_THREAD_MEDIUM_STACKSIZE];
+static char WORD_ALIGNED _evq_lowest_stack[EVENT_THREAD_LOWEST_STACKSIZE];
 
-#ifdef MODULE_EVENT_THREAD_LOWEST
-event_queue_t event_queue_lowest;
-static char _evq_lowest_stack[EVENT_THREAD_LOWEST_STACKSIZE];
-#endif
+event_queue_t event_thread_queues[EVENT_QUEUE_PRIO_NUMOF];
 
-typedef struct {
-    event_queue_t *queue;
-    char *stack;
-    size_t stack_size;
-    unsigned priority;
-} event_threads_t;
-
-const event_threads_t _event_threads[] = {
-#ifdef MODULE_EVENT_THREAD_HIGHEST
-    { &event_queue_highest, _evq_highest_stack, sizeof(_evq_highest_stack),
-        EVENT_THREAD_HIGHEST_PRIO },
-#endif
-#ifdef MODULE_EVENT_THREAD_MEDIUM
-    { &event_queue_medium, _evq_medium_stack, sizeof(_evq_medium_stack),
-        EVENT_THREAD_MEDIUM_PRIO },
-#endif
-#ifdef MODULE_EVENT_THREAD_LOWEST
-    { &event_queue_lowest, _evq_lowest_stack, sizeof(_evq_lowest_stack),
-        EVENT_THREAD_LOWEST_PRIO },
-#endif
-};
 
 void auto_init_event_thread(void)
 {
-    for (unsigned i = 0; i < ARRAY_SIZE(_event_threads); i++) {
-        event_thread_init(_event_threads[i].queue,
-                _event_threads[i].stack, _event_threads[i].stack_size,
-                _event_threads[i].priority);
+    if (IS_USED(MODULE_EVENT_THREAD_HIGHEST)) {
+        event_thread_init(EVENT_PRIO_HIGHEST,
+                          _evq_highest_stack, sizeof(_evq_highest_stack),
+                          EVENT_THREAD_HIGHEST_PRIO);
+    }
+    if (IS_USED(MODULE_EVENT_THREAD_MEDIUM)) {
```

Comment:
```suggestion
    if (IS_USED(MODULE_EVENT_THREAD_LOWEST)) {
```

Reply:
This is actually correct. The medium priority thread is always started, only the lower priority and the higher priority thread gets spawned on demand.

The idea is that module `event_thread_highest` allows the highest priority events to preempt all others. For this, we need to spawn the highest priority thread. The `event_thread_medium` module allows medium priority events to preempt low priority events. For that, we need to spawn the low priority thread.

This is not super ideal modeled, but this way the preexisting modules provide the same real time guarantees es before. So this gives us backward compatibility. I think on the long term this should be configured by other means than modules, but let's wait for reign of KConfig to begin first.

Reply:
Huh, but on the next line a thread with prio `EVENT_THREAD_LOWEST_PRIO` is spawned

Reply:
Exactly :-) This is how it should be

Reply:
Ah ok so if I only want one thread, I must neither select `MODULE_EVENT_THREAD_MEDIUM` nor `MODULE_EVENT_THREAD_HIGHEST` - and there is no such thing as `MODULE_EVENT_THREAD_LOWEST`.

Please add a comment, at the first glance this looks like a typo

Reply:
I added some comments to explain the reasoning behind spawning the low priority thread when the medium priority module is used.